WeScan/Edit/EditImageViewController.swift

```diff
@@ -0,0 +1,165 @@
+        imageView.isOpaque = true
+        imageView.image = image
+        imageView.backgroundColor = .black
+        imageView.contentMode = .scaleAspectFit
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
+    
+    lazy private var quadView: QuadrilateralView = {
+        let quadView = QuadrilateralView()
+        quadView.editable = true
+        quadView.translatesAutoresizingMaskIntoConstraints = false
+        return quadView
+    }()
+    
+    // MARK: - Life Cycle
+    
+    public init(image: UIImage, quad: Quadrilateral?, rotateImage: Bool = true) {
+        self.image = rotateImage ? image.applyingPortraitOrientation() : image
+        self.quad = quad ?? EditImageViewController.defaultQuad(forImage: image)
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override public func viewDidLoad() {
+        super.viewDidLoad()
+        
+        setupViews()
+        setupConstraints()
+        zoomGestureController = ZoomGestureController(image: image, quadView: quadView)
+        
+        let touchDown = UILongPressGestureRecognizer(target: zoomGestureController,
+                                                     action: #selector(zoomGestureController.handle(pan:)))
+        touchDown.minimumPressDuration = 0
+        view.addGestureRecognizer(touchDown)
+    }
+    
+    override public func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+        adjustQuadViewConstraints()
+        displayQuad()
+    }
+    
+    // MARK: - Setups
+    
+    private func setupViews() {
+        view.addSubview(imageView)
+        view.addSubview(quadView)
+    }
+    
+    private func setupConstraints() {
+        let imageViewConstraints = [
+            imageView.topAnchor.constraint(equalTo: view.topAnchor),
+            imageView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            view.bottomAnchor.constraint(equalTo: imageView.bottomAnchor),
+            view.leadingAnchor.constraint(equalTo: imageView.leadingAnchor)
+        ]
+        
+        quadViewWidthConstraint = quadView.widthAnchor.constraint(equalToConstant: 0.0)
+        quadViewHeightConstraint = quadView.heightAnchor.constraint(equalToConstant: 0.0)
+        
+        let quadViewConstraints = [
+            quadView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            quadView.centerYAnchor.constraint(equalTo: view.centerYAnchor),
+            quadViewWidthConstraint,
+            quadViewHeightConstraint
+        ]
+        
+        NSLayoutConstraint.activate(quadViewConstraints + imageViewConstraints)
+    }
+    
+    // MARK: - Actions
+    public func cropImage(){
+        guard let quad = quadView.quad, let ciImage = CIImage(image: image) else {
+            delegate?.cropped(image: nil)
```

Comment:
Why are we calling `cropped` here while we didn't crop anything? Shouldn't we delegate with a failure callback instead?

Reply:
I agree about image param should not be optional.